measurements/pages/__init__.py

```diff
@@ -204,34 +208,69 @@ def files_in_country(country_code):
                            order_by=order_by,
                            current_country=country_code)
 
-
-@pages_blueprint.route('/files/download/<filename>')
-def files_download(filename):
-    try:
-        report = current_app.db_session.query(Report) \
-                        .filter(Report.textname.like("%" + filename)).first()
-        # XXX
-        # We have duplicate measurements :(
-        # So the below exception actually happens. This should be resolved
-        # in the data processing pipeline.
-        #   Report.filename == filename).one()
-    except NoResultFound:
-        raise NotFound("No file with that filename found")
-    except MultipleResultsFound:
-        # This should never happen.
-        raise HTTPException("Duplicate measurement detected")
-
-    if report is None:
+def decompress_autoclaved(autoclaved_filename, frames_off, frames_size, intra_off, intra_size):
+    def generator():
+        url = urljoin(current_app.config['AUTOCLAVED_BASE_URL'], autoclaved_filename)
+
+        current_app.logger.debug("Fetching autoclaved from: %s" % url)
+        current_app.logger.debug("  frames_off: %d" % frames_off)
+        current_app.logger.debug("  frames_size: %d" % frames_size)
+        current_app.logger.debug("  intra_off: %d" % intra_off)
+        current_app.logger.debug("  intra_size: %d" % intra_size)
+        try:
+            r = requests.get(url, headers={"Range": "bytes={}-{}".format(frames_off, frames_off + frames_size)},
+                                stream=True)
+            streamed_data = 0
+            for chunk in lz4framed.Decompressor(r.raw):
+                current_app.logger.debug("chunking away 1")
+                d = chunk
+                current_app.logger.debug("chunking away 2")
+                if streamed_data == 0 and intra_off > 0:
+                    d = chunk[intra_off:]
+                if streamed_data > (intra_size + len(d)):
+                    d = chunk[:(streamed_data - intra_size)]
```

Comment:
Have you tested it? It looks a bit fishy, but I'm unsure if I understand the logic correctly.

Reply:
It would help, if not to have integration tests, at least to factor out this logic into unit test cases and see whether it works in such cases or not. (It may also help a comment showing the general case and the pointers.)

Reply:
I think that taking some files from the pipeline (or downloading them on demand) is a reasonable way to go.
IMHO, there are several corner-cases to test:
- report fitting into single frame
- report spanning 2 frames
- report spanning >2 frames (having at least one full-sized frame)
- report aligned with frame beginning
- report aligned with frame end

Reply:
@darkk do you have an easy way of generating reports that fit these criteria? That would make it easier to write some test cases for this.

Reply:
I'd suggest to seek for existing blobs in the database, tell me if you need help with that. IMHO that would make test way more end-to-end.

Reply:
@darkk wrote:
> IMHO, there are several corner-cases to test:
>
> - report fitting into single frame
> - report spanning 2 frames
> - report spanning >2 frames (having at least one full-sized frame)
> - report aligned with frame beginning
> - report aligned with frame end

Yes, I cannot think of more corner cases to test.